### ISSUE TITLE
[BUG] CQL fails to converge

### ISSUE THREAD

**Issue by pietroorlandi** [takuseno/d3rlpy#457]
CQL does not converge when trained on several custom datasets. Even with conservative settings, the critic loss does not stabilize: it tends to oscillate for a while and then increases significantly, showing divergence instead of convergence.

**To Reproduce**
Run the following code:
```
import d3rlpy
from d3rlpy.algos import CQLConfig

# Replace this with your own MDPDataset
# dataset = MDPDataset(observations=..., actions=..., rewards=..., terminals=...)

# For the sake of minimal reproduction, here's an example with a built-in dataset:
dataset, env = d3rlpy.datasets.get_dataset('pendulum-random')

cql_very_stable = CQLConfig(
    actor_learning_rate=1e-5,
    critic_learning_rate=1e-5,
    batch_size=512,
    n_critics=2,
    initial_alpha=0.1,
    alpha_threshold=1.0,
    conservative_weight=1.0,
    n_action_samples=10,
    soft_q_backup=True,
    compile_graph=True,
).create(device='cuda:0')

cql_very_stable.fit(
    dataset,
    n_steps=100000,
    show_progress=True
)
```
**Expected behavior**
I expected the critic loss to gradually decrease or remain stable, indicating training progress. Instead, it oscillates and often grows rapidly after a few thousand steps.

**Additional context**

- d3rlpy version: 2.8.1
- torch version: 2.7.1
- CUDA version: 12.5

The issue also occurs with my own custom datasets (created with MDPDataset), not just the built-in ones.

I haven't performed extensive hyperparameter tuning, but even very conservative settings (low learning rates, high conservative weight, etc.) seem ineffective in stabilizing the training.

Please let me know if further details (e.g., logs or plots) would be helpful. Thank you for your work on this excellent project!

**Comment by pietroorlandi** [comment]
@takuseno Thanks for the suggestion. I tried running the same configuration as before, but with alpha_learning_rate=0 to disable automatic alpha tuning. Unfortunately, the results seem even worse: the critic loss increases rapidly and doesn’t stabilize at all.

My ultimate goal is to train a reinforcement learning agent on a custom offline dataset I collected. Do you have any recommendations for more stable algorithms for this kind of use case, or do most offline RL methods suffer from similar convergence issues?

**Comment by takuseno** [comment]
Sounds like you need to use `MinMaxActionScaler` as follows:
https://github.com/takuseno/d3rlpy/blob/4f0956ba8db469ebeeb617bb28060a8401a7dfa6/examples/preprocessors.py#L32

This will scale action outputs to a range within your dataset action value range. Currently, your action outputs are scaled within [-1, 1].

**Comment by pietroorlandi** [comment]
@takuseno Thanks! I was already using `MinMaxActionScaler`, as well as `MinMaxObservationScaler` and `MinMaxRewardScaler`, but the divergence issue still persists.

**Comment by hc1991** [comment]
I encountered the same problem，
My config is:

<img width="518" height="246" alt="Image" src="https://github.com/user-attachments/assets/c3c8893f-d439-4192-a07e-d9cba8267805" />

1、the std loss does not stabilize: it tends to oscillate for a while and then increases significantly, showing divergence instead of convergence.
2、When the number of training times reaches the value set by target_update_interval, the loss becomes larger and then oscillate. 

<img width="781" height="398" alt="Image" src="https://github.com/user-attachments/assets/3a529e1d-cb69-43d8-9611-ac8fd1cba91f" />


target_update_interval is 4000，

3、Same behavior when I use the example you gave，example：https://colab.research.google.com/github/takuseno/d3rlpy/blob/master/tutorials/cartpole.ipynb

<img width="1039" height="458" alt="Image" src="https://github.com/user-attachments/assets/0a7b1c03-a379-46dc-9dfa-2cc65ddf1555" />

@takuseno

**Comment by takuseno** [comment]
@hc1991 Thank you for reporting the issue. In offline RL, it's not guaranteed that the loss is converged. If the agent performs well in evaluation episodes, it's not really a problem.

**Comment by hc1991** [comment]
@takuseno  thanks for your reply.

I am a newbie who is new to reinforcement learning. I use the d3rlpy framework to learn. I evaluated the effect of the model as you said,

evaluators = {
        'initial_Evaluator':InitialStateValueEstimationEvaluator(test_data),
        'soft_opc_Evaluator': SoftOPCEvaluator(return_threshold=return_75,episodes=test_data.episodes),
} 

My config:
batchsize=128,alpha=2.0,learning_rate=2e-5,

I found 3 problems:

1. I use SoftOPC for evaluation. The data performance is that it fluctuates in the first few rounds and then rises steadily. 
   My return_threshold parameter is the 75% quantile，SoftOPC’s final score is 0.05  (50 epoch).

```python
   episode_returns = []
    for i, episode in enumerate(dataset.episodes):
        total_return = episode.compute_return()
        episode_returns.append({'index': i, 'total_return': total_return})
    returns_df = pd.DataFrame(episode_returns)
    print(returns_df['total_return'].describe())

    return_threshold = returns_df['total_return'].quantile(0.75)
  ```
   
   
2. I user InitialStateValueEstimationEvaluator,It has a bug, the observations passed cannot be Sequence[NDArray].
   Because the internal implementation will only take the observations first element, My observations is a tuple containing 2 array
   
```python
class InitialStateValueEstimationEvaluator:
    # estimate action-value in initial states
    first_obs = np.expand_dims(batch.observations[0], axis=0)
```


   
3. I use the final model to do a simple test on my data

```python

algo = d3l.load_learnable(fname="model_10000.d3")
result = algo.predict(observations)
pd.Series(result).value_counts(normalize=True)
```

output:
0    0.999789
1    0.000106
2    0.000102
3    0.000003

The 99% probability is the selected action 0


My action and reward data distribution：

=== Summary of Rewards distribution ===

| action | count     | mean      | std       | min        | 25%       | 50%       | 75%       | max        |
|--------|-----------|-----------|-----------|------------|-----------|-----------|-----------|------------|
| 0      | 245923.0  | 9.672059  | 4.778022  | -10.000000 | 9.345417  | 10.310900 | 11.662549 | 16.703172  |
| 1      | 162466.0  | 10.108939 | 4.808394  | -10.000000 | 9.724581  | 10.857436 | 12.136153 | 16.703172  |
| 2      | 101453.0  | 10.059513 | 5.285181  | -9.960397  | 9.787598  | 11.121330 | 12.350218 | 16.703172  |
| 3      | 153813.0  | 10.132998 | 5.598752  | -10.000000 | 9.841762  | 11.371911 | 12.658355 | 16.703172  |
| 4      | 60336.0   | 10.189240 | 6.157496  | -10.000000 | 9.950498  | 11.733509 | 13.275175 | 16.703172  |


=== Summary of Episode Returns ===
count    185690.000000
mean         38.895669
std          39.066757
min        -120.485809
25%          12.236826
50%          26.078649
75%          52.656541
max         403.430542
Name: total_return, dtype: float64

High-return threshold (75th percentile): 52.66
Number of high-return episodes(75th percentile): 46423

=== Action Frequency in ALL Episodes ===
0    0.339677
1    0.224403
2    0.140130
3    0.212452
4    0.083338
Name: proportion, dtype: float64

=== Action Frequency in HIGH-RETURN((75th percentile)) Episodes ===
0    0.384938
1    0.241166
2    0.139463
3    0.182309
4    0.052125
Name: proportion, dtype: float64

=== Action Distribution Comparison (Validation Set) ===
| action | All Trajectories (%) | High-Return (75th percentile) Trajectories (%) | Enrichment Factor |
|--------|-----------------------|----------------------------------------------|--------------------|
| 0      | 33.97                 | 38.49                                        | 1.13               |
| 1      | 22.44                 | 24.12                                        | 1.07               |
| 2      | 14.01                 | 13.95                                        | 1.00               |
| 3      | 21.25                 | 18.23                                        | 0.86               |
| 4      |  8.33                 |  5.21                                        | 0.63               |

       


The training results look bad
I tried adjusting parameters such as alpha, learning_rate, batch size, etc. The performance is still the same,Can you provide some suggestions to help me troubleshoot possible issues?
